src/coreclr/nativeaot/Common/src/Internal/Runtime/MethodTable.cs

```diff
@@ -732,7 +732,14 @@ internal bool RequiresAlign8
         {
             get
             {
-                return (RareFlags & EETypeRareFlags.RequiresAlign8Flag) != 0;
+                if (HasComponentSize)
+                {
+                    return (RareFlags & EETypeRareFlags.RequiresAlign8Flag) != 0;
+                }
+                else
+                {
+                    return (ExtendedFlags & (ushort)EETypeFlagsEx.RequiresAlign8Flag) != 0;
+                }
```

Comment:
Why duplicate this information into `EETypeFlagsEx`?

Reply:
I was just writing a comment on that...

`GCStaticEETypeNode` uses a truncated method table that is only meant to be interpreted for the GC allocation purposes. In order to emit the current definition of `RequiresAlign8Flag` we would need to emit all the missing fields, the optional fields node and pointer to it. That's quite expensive thing to do. It requires more code in the compiler, more data in the object file, and more decoding at runtime. The idea is to move the flag to extended flags for non-array/non-string types since there's plenty of unused bits. The value `0x1000` is specifically chosen to match `MTFlag_RequiresAlign8` in GC headers which was previously unused on NativeAOT.

Side effect is that type allocation is slightly cheaper for the cases where we don't need to decode the optional fields, and in some cases we may save some object file space too.

Reply:
It makes the check for RequiresAlign8 more expensive at runtime though and we don't currently have any duplicated flags.

Reply:
Do we think the "allocate every threadstatic as Align8" would have meaninigfully bad perf impact?

Reply:
One could also in theory just store the requiresalign8 bit in the lowest bit of the GCStatic EEType pointer and check for that in AllocateThreadStaticStorageForType. (Just throwing stuff around because I'm not a fan of the duplication of flags and don't like the idea of having to refactor GCStatic EEType emission to be shared with regular EETypes so that we can emit rare flags properly)

Reply:
I doesn’t really make it more expensive. For `RareFlags` you have to check `Flags & HasOptionalFields`. For the pessimistic case this can still be folded to essentially the same number of AND/comparison instructions. For the optimistic case it saves decoding the optional fields completely, which is relatively expensive operation.

This is happening for every object allocation at runtime.

Reply:
The decision to put RequiresAlign8 into Rare flags was done when Arm32 was still a relevant mobile platform. It was likely done that way because almost no type has rare flags set and almost no type RequiresAlign8. The 99% case was probably just a check whether there's rare flags.

Reply:
> Do we think the "allocate every threadstatic as Align8" would have meaninigfully bad perf impact?

Not necessarily. I checked the tests and there are basically no thread static data that would require Align8 at the moment, so at minimum it would be a small memory waste.

Reply:
> One could also in theory just store the requiresalign8 bit in the lowest bit of the GCStatic EEType pointer…

Funnily enough I originally used this strategy when I prototyped it. That still gets us into situation where the “Align8” flag is stored in two different ways. I am not entirely opposed to it but I eventually went with a strategy that seemed to have benefits for other cases while keeping single source of truth (the `MethodTable.RequiresAlign8` getter).

Reply:
> The 99% case was probably just a check whether there's rare flags and the answer was no.

`RareFlags` are in the optional properties block. In order to check it you need to decode the block and read the flags even if, by your own admission, in 99% case this will be wasted time. The check happens for every allocation in `RhNewObject`.

What this change does is that for anything non-array/non-string the cost goes significantly down since you are just checking the low bits of the `Flags` field that you are already accessing.

Now the `Align8` in `RareFlags` is only ever used for something like `double[]` and you only pay the decoding cost of the optional fields for arrays/strings and only if `HasOptionalFields` is set.

Reply:
> `RareFlags` are in the optional properties block. In order to check it you need to decode the block and read the flags even if, by your own admission, in 99% case this will be wasted time.

The optional fields are rare as well so most of the time we're done after a failed check for optional fields presence.

I agree this change would make some of the rare cases faster at the cost of making the common case slower (and adding new complexity dimension to MethodTable data structure). Combining a correctness fix and a perf fix is always problematic (perf fixes typically come with benchmark numbers justifying them).

Reply:
I compiled `nativeaot/SmokeTests/UnitTests` to get some data. There are 14471 `EETypeNode`s in total. 2310 of those encode optional fields (~16%). 166 require `Align8`, of those 17 are arrays. Obviously it doesn't say anything about the actual usage of the types, but for 149 of them we would now use the faster `ExtendedFlags` code path and this includes types like `DateTime`, `TimeSpan`, `DateTimeOffset`, `Decimal`, boxed `[U]Int64`, boxed `double` and many types used in thread pool accounting. I really don't see how we can end up with a worse performance. It's one more AND of a value that is already in a register. The `RequiresAlign8` check is only done on ARM and WASM so it doesn't affect other platforms at all.

Think of it as
```
var regFlags = Flags;
if ((regFlags & (HasComponentSize | Align8)) == Align8)
  return true;
if ((regFlags & (HasComponentSize | HasOptionalFields) == (HasComponentSize | HasOptionalFields))
  return RareFlags & Align8;
return false; 
```

> ...and adding new complexity dimension to MethodTable data structure

That's fair, and I agree. I just think the complexity is really limited to two place - the one that emits `EETypeNode`, and the `RequiresAlign8` getter. This doesn't need changes anywhere else and doesn't leak into zillion of places that would need to be changed.

Reply:
We currently set rare flags on a lot more types than needed. #105945 fixes that.

From the number of types that encode optional fields, we should also remove nullables and byref-like types. Neither of them would ever meaningfully show up in a newobj.

Reply:
> From the number of types that encode optional fields, we should also remove nullables and byref-like types.

50 of the reported types are `Nullable`. I didn't dump the exact info on byref-like, 231 of the types are `Span`/`ReadOnlySpan` so that may serve like a very rough estimate. That's still around 14% of all EEType (but since those are not going to be allocated we should also subtract them from the divisior). I can rerun the numbers later with the changes #105945 to get a more precise answer. It would probably be possible to add counters to the object allocation paths to get the numbers for an actual run.

Reply:
I would note that it is possible to collapse the duplicate flags at the cost of making the getter a bit slower still:
```
if (!HasComponentSize())
    return ExtendedFlags & Align8;
else
    return ElementType->IsValueType && ElementType->IsAlign8;
```

Reply:
@SingleAccretion I will try to prototype it in separate branch tomorrow. I think it can be quite clean if we move some of the checks one level up and change `RequiresAlign8` to be valid only for value types (w/ assert).